peps/pep-0788.rst

```diff
@@ -224,254 +287,290 @@ created by the authors of this PEP:
-but is still limited by the current API.
+explicit interpreter, rather than assuming that the main interpreter was
+requested), but is still limited by the current hanging problems in the C API.
 
-In particular, subinterpreters typically have a much shorter lifetime than the
-main interpreter, and as such, there's not necessarily a guarantee that a
-:c:type:`PyInterpreterState` (acquired by :c:func:`PyInterpreterState_Get`)
-passed to a fresh thread will still be alive. Similarly, a
-:c:type:`PyInterpreterState` pointer could have been replaced with a *new*
-interpreter, causing all sorts of unknown issues. They are also subject to
-all the finalization related hanging mentioned previously.
+In addition, subinterpreters typically have a much shorter lifetime than the
+main interpreter, so there's a much higher chance that an interpreter passed
+to a thread will have already finished and have been deallocated. So, passing
+that interpreter to :c:func:`PyThreadState_New` will most likely crash the program
+because of a use-after-free on the interpreter-state.
 
 Rationale
 =========
 
-This PEP includes several new APIs that intend to fix all of the issues stated
-above.
+So, how do we address all of this? The best way seems to be starting from
+scratch and "reimagining" how to create, acquire and attach
+:term:`thread states <thread state>` in the C API.
+
+Preventing Interpreter Shutdown with Reference Counting
+-------------------------------------------------------
 
-Replacing the old APIs
-----------------------
+This PEP takes an approach where an interpreter is given a reference count
+that prevents it from shutting down.
 
-As made clear in Motivation_, ``PyGILState`` is already pretty buggy, and
-even if it was magically fixed, the current behavior of hanging the thread is
-beyond repair. In turn, this PEP intends to completely deprecate the existing
-``PyGILState`` APIs and provide better alternatives. However, even if this PEP
-is rejected, all of the APIs can be replaced with more correct ``PyThreadState``
-functions in the current C API:
+So, holding a "strong reference" to the interpreter will make it safe to
+call the C API without worrying about the thread being hung.
+
+This means that interfacing Python (for example, in a C++ library) will need
+a reference to the interpreter in order to safely call the object, which is
+definitely more inconvenient than assuming the main interpreter is the right
+choice, but there's not really another option.
+
+Weak References
+***************
+
+This proposal also comes with weak references to an interpreter that don't
+prevent it from shutting down, but can be promoted to a strong reference when
+the user decides that they want to call the C API. Promotion of a weak reference
+to a strong reference can fail if the interpreter has already finalized, or
+reached a point during finalization where it can't be guaranteed that the
+thread won't hang.
+
+Removing the GIL-state APIs
```

Comment:
I suggest to rename it to "Deprecating ..." since the API will stay for many years.